File: backend/providers/geckoterminal.py
import os
import requests
import logging

logger = logging.getLogger("tokencare-geckoterminal")
REQUEST_TIMEOUT = 4.0

def fetch_geckoterminal_token(contract_address: str, geckoterminal_network: str) -> dict:
    """
    Fetches token data from GeckoTerminal API
    """
    api_key = os.getenv("GECKOTERMINAL_API_KEY", "")
    headers = {"Accept": "application/json;version=20230302"}
    if api_key:
        headers["Authorization"] = f"Bearer {api_key}"

    try:
        url = f"https://api.geckoterminal.com/api/v2/networks/{geckoterminal_network}/tokens/{contract_address.lower()}"
        resp = requests.get(url, headers=headers, timeout=REQUEST_TIMEOUT)
        if resp.status_code == 200:
            data = resp.json()
            attr = data.get("data", {}).get("attributes", {})
            if attr:
                price_usd = float(attr.get("price_usd") or 0.0)
                volume_24h = float((attr.get("volume_usd") or {}).get("h24") or 0.0)
                liquidity = float(attr.get("total_reserve_in_usd") or 0.0)
                market_cap = float(attr.get("fdv_usd") or 0.0)
                logo = attr.get("image_url") or None

                return {
                    "source": "geckoterminal",
                    "found": True,
                    "name": attr.get("name"),
                    "symbol": str(attr.get("symbol") or "").upper(),
                    "logo": logo,
                    "price": price_usd,
                    "priceUsd": price_usd,
                    "priceChange24h": 0.0,
                    "volume24h": volume_24h,
                    "liquidity": liquidity,
                    "marketCap": market_cap,
                    "verified": True
                }
    except Exception as e:
        logger.warning(f"[GeckoTerminal] Error fetching {contract_address}: {e}")

    return {"source": "geckoterminal", "found": False}

File: backend/providers/geckoterminal.ts
import { NormalizedTokenDetails } from '../types/tokenDetails';

export const CHAIN_MAP_GECKOTERMINAL: Record<string, string> = {
  '1': 'eth',
  '137': 'polygon_pos',
  '56': 'bsc',
  '42161': 'arbitrum',
  '10': 'optimism',
  '8453': 'base',
  '43114': 'avax',
  'solana': 'solana',
  'ethereum': 'eth',
  'polygon': 'polygon_pos',
};

export async function fetchGeckoTerminalToken(
  chain: string,
  contractAddress: string
): Promise<NormalizedTokenDetails | null> {
  const cleanAddr = contractAddress.trim().toLowerCase();
  if (!cleanAddr) return null;

  try {
    const network = CHAIN_MAP_GECKOTERMINAL[chain.toLowerCase()] || 'eth';
    const url = `https://api.geckoterminal.com/api/v2/networks/${network}/tokens/${cleanAddr}`;

    const res = await fetch(url, {
      headers: { 'Accept': 'application/json' },
    });

    if (!res.ok) return null;
    const json = await res.json();
    const data = json.data?.attributes;
    if (!data) return null;

    const isSolana = network === 'solana';

    const normalized: NormalizedTokenDetails = {
      name: data.name || 'Unknown Token',
      symbol: (data.symbol || 'TOKEN').toUpperCase(),
      decimals: Number(data.decimals) || (isSolana ? 9 : 18),
      contractAddress: cleanAddr,
      chain: chain,
      chainId: chain,
      blockchain: isSolana ? 'Solana' : 'EVM',
      assetStandard: isSolana ? 'SPL' : 'ERC20',
      logoUrl: data.image_url || undefined,
      priceUsd: Number(data.price_usd) || undefined,
      fdvUsd: Number(data.fdv_usd) || undefined,
      marketCapUsd: Number(data.market_cap_usd) || Number(data.fdv_usd) || undefined,
      volume24hUsd: Number(data.volume_usd?.h24) || 0,
      totalSupply: data.total_supply,
      resolvedVia: 'geckoterminal',
    };

    return normalized;
  } catch (err) {
    console.error('[GeckoTerminal Provider Error]:', err);
    return null;
  }
}
